Make all code async

dimmy-timmy opened the pull request:
Current code is mostly sync which could lead to thread pool starvation for highload scenarios 
We have this issue currently with redis . We are observing a lot of errors like `Timeout performing EVAL`

https://github.com/ikyriak/IdempotentAPI/issues/23

ikyriak commented:
Hello @dimmy-timmy,

Thank you for taking the time to implement this and for sharing it 🙏

ikyriak commented on src/IdempotentAPI/Filters/IdempotencyAttributeFilter.cs:
```diff
@@ -49,14 +50,16 @@ public class IdempotencyAttributeFilter : IActionFilter, IResultFilter
             // If the Idempotency is disabled then stop
             if (!_enabled)
             {
+                await next();
```
Thanks @dimmy-timmy. Nice catch💪

ikyriak commented on src/IdempotentAPI/Filters/IdempotencyAttributeFilter.cs:
```diff
@@ -72,54 +75,36 @@ public void OnActionExecuting(ActionExecutingContext context)
+            await _idempotency.ApplyPreIdempotency(context);
+            
+            var result = await next();
+            if (result?.Exception is not null)
             {
-                _idempotency.CancelIdempotency();
+                await _idempotency.CancelIdempotency();
             }
```
The `OnActionExecuted` applies the Pre-idempotency and, in conditions, returns the cached response based on the provided idempotencyKey. So, I was thinking of executing the `next()` only when the `_isPreIdempotencyCacheReturned` is False (i.e., when we do not have something to return by exposing it as read-only property). Otherwise, the following filter may alter our predefined cached response. Wdyt?

ikyriak replied:
Yes, that's it. Thank you for the information 😄

dimmy-timmy commented:
@ikyriak I also added github actions configuration to run ci build and test on pull request.
And modified Integration Tests to be able to run them on CI using `WebApplicationFactory`

dimmy-timmy commented on src/IdempotentAPI.AccessCache/Lockers/InProcessAccessLock.cs:
```diff
@@ -15,44 +17,53 @@ namespace IdempotentAPI.AccessCache.Lockers
-                {
-                    disposeTracker = new LockDisposeTracker();
-                    lockValueMapper.Add(valueToLock, disposeTracker);
-                }
-                disposeTracker.Count++;
-            }
-            Monitor.Enter(disposeTracker);
+                Count = disposeTracker.Count + 1
+            });
+            await disposeTracker.Semaphore.Value.WaitAsync();
+            return new InProcessAccessLock(valueToLock);
```
It was needed to rewrite in-proc lock to use `SemaphoreSlim` because `Monitor` does not support async flow

ikyriak replied:
Thank you, @dimmy-timmy for your effort ❤️

ikyriak replied:
Modifying the integration tests to use WebApplicationFactory is an excellent addition 💪 . I am removing it from my TODO list 😄

ikyriak commented on src/IdempotentAPI.AccessCache/Lockers/InProcessAccessLock.cs:
```diff
@@ -15,44 +17,53 @@ namespace IdempotentAPI.AccessCache.Lockers
         public void Dispose()
         {
-            lock (localLocker)
+            while (true)
```
Why is there a need for a `while (true)` here?

dimmy-timmy replied:
@ikyriak  To ensure that value was updated in concurrent dictionary
basically the flow is following 
loop until it get updated
`TryUpdate` accepts original value and updated value and return false if original value was changed from other thread

dimmy-timmy replied:
@ikyriak Hi, do you have any other questions ?

ikyriak replied:
@dimmy-timmy Thank you for your response. I will include your improvements in the `2.1.0-rc-01` release.

karena-st commented:
Hey guys, I was also about to propose this change to get rid of thread starvation. Luckily I saw the PR already opened.
@ikyriak when do you think we can have it in RC ? to be able to use it as a package ...

ikyriak commented:
@dimmy-timmy Thank you again for your time and effort 🙏.

@karena-st 
The `async` improvements have been released in [v2.1.0](https://github.com/ikyriak/IdempotentAPI/pull/54).
